crowbar_framework/app/models/api/crowbar.rb

```diff
@@ -80,6 +80,60 @@ def maintenance_updates_installed?
       end
     end
 
+    def compute_resources_available?
+      compute_nodes = NodeObject.find("roles:nova-compute-kvm")
+      return true if compute_nodes.empty?
+      if compute_nodes.size == 1
+        Rails.logger.warn("Only one compute node found; non-disruptive upgrade is not possible!")
+        return false
+      end
+
+      # Now, find out if we have enough memory for migrating
+
+      # Find the controller responsible for the nova services
+      founders = NodeObject.find(
+        "roles:nova-controller AND pacemaker_founder:true AND pacemaker_config_environment:*"
+      )
+      controller = founders.first
+
+      free_compute_memory = 0
+      largest_compute_node_total = 0
+      largest_compute_node_free = 0
+
+      compute_nodes.each do |compute_node|
+        # FIXME: are we sure nova uses the short name of node for its hosts?
+        name = compute_node.handle
+
+        total_cmd = ". /root/.openrc; /usr/bin/nova host-describe #{name} | grep '(total)'"
+        ssh_retval = controller.run_ssh_cmd(total_cmd)
+        out = ssh_retval[:stdout] || ""
+        total_memory = out.split[7].to_i
+
+        used_cmd = ". /root/.openrc; /usr/bin/nova host-describe #{name} | grep '(used_now)'"
+        ssh_retval = controller.run_ssh_cmd(used_cmd)
+        out = ssh_retval[:stdout] || ""
+        used_memory = out.split[7].to_i
+
+        free_memory = total_memory - used_memory
+        free_compute_memory += free_memory
+
+        # find the node with largest total memory
+        if total_memory > largest_compute_node_memory
+          largest_compute_node_total = total_memory
+          largest_compute_node_free = free_memory
+        end
+      end
+
+      if free_compute_memory - largest_compute_node_free < largest_compute_node_total
```

Comment:
@nicolasbock The motivation behind this logic from my side was to think about a solution which incorporates the cluster at different points in scale in and out, while trying to cater a simple and easy solution especially with limited feedback channels for getting the right data to make these decisions. The major motivation from my side was to consider if the `Free Memory of the Compute Cluster` is more than the given node which needs to be free'd from all the VM's. Now if this given Compute Node is the largest, then we should consider the worst case where this largest compute node is actually 100% provisioned (let's assume that I am more crazier than nova-scheduler), hence we should have this considered: `Free Memory of the Compute Cluster` >= `Memory of the Largest Compute Node`, but the largest compute node is going to be removed from the cluster! Hence: (`Free Memory of the Compute Cluster` - `Free Memory of the Largest Compute Node`) >= `Free Memory of the Largest Compute Node`. We do not have to migrate all the instances on one compute node. We just need to make sure that the largest compute node could be easily emptied. This logic should be applicable for a 2 node cluster and also for a 1000 node cluster, although I do not know if it is as efficient for a 1000 node cluster especially considering the network traffic generated by the queries for the same and all that jazz. I would be really happy if you could disprove me here because I cannot believe it myself that I could cook this up in a few minutes.

I wanted to make this a non-linear equation but my mathematics is rusty or rather, CPU's could be over-provisioned much better so counting their factor should be limiting to zero (for modern decently powerful server CPU farms ... remember? we have a cluster!) and the hard disk should ideally be on shared storage. So this algorithm although not quite exhaustive is still not missing out anything critical. Thanks for trying to understand this piece of code.